python/ray/dag/compiled_dag_node.py

```diff
@@ -1075,6 +1192,202 @@ def _get_or_compile(
+                )
+        return actor_to_operation_nodes
+
+    def _build_dag_node_operation_graph(
+        self,
+        actor_to_operation_nodes: Dict[
+            "ray.actor.ActorHandle", List[List[DAGOperationGraphNode]]
+        ],
+    ) -> Dict[int, Dict[DAGNodeOperationType, DAGOperationGraphNode]]:
+        """
+        Generate a DAG node operation graph by adding edges based on the
+        following rules:
+
+        #1  Add edges from READ to COMPUTE, and from COMPUTE to WRITE, which
+            belong to the same task.
+        #2  Add an edge from COMPUTE with bind_index i to COMPUTE with bind_index
+            i+1 if they belong to the same actor.
+        #3  Add an edge from WRITE of the writer task to READ of the reader task.
+
+        This is the step one of building an execution schedule for each actor.
+
+        Args:
+            actor_to_operation_nodes: A dictionary that maps an actor handle to
+                a list of lists of DAGOperationGraphNode. For the same actor, the
+                index of the outer list corresponds to the index of the ExecutableTask
+                in the list of `executable_tasks` in `actor_to_executable_tasks`. In
+                the inner list, the order of operations is READ, COMPUTE, and WRITE.
+
+        Returns:
+            A graph where each node is a DAGOperationGraphNode. The key is the index
+            of the task in idx_to_task, and the value is a dictionary that maps the
+            DAGNodeOperationType (READ, COMPUTE, or WRITE) to the corresponding
+            DAGOperationGraphNode.
+        """
+        assert self.idx_to_task
+
+        graph: Dict[int, Dict[DAGNodeOperationType, DAGOperationGraphNode]] = {}
+
+        for _, operation_nodes_list in actor_to_operation_nodes.items():
+            prev_compute_node = None
+            for operation_nodes in operation_nodes_list:
+                idx = operation_nodes[0].idx
+                read_node, compute_node, write_node = (
+                    operation_nodes[0],
+                    operation_nodes[1],
+                    operation_nodes[2],
+                )
+                # Add edges from READ to COMPUTE, and from COMPUTE to WRITE, which
+                # belong to the same task.
+                read_node.add_edge(compute_node)
+                compute_node.add_edge(write_node)
+                # Add an edge from COMPUTE with `bind_index` i to COMPUTE with
+                # `bind_index` i+1 if they belong to the same actor.
+                if prev_compute_node is not None:
+                    prev_compute_node.add_edge(compute_node)
+                prev_compute_node = compute_node
+                assert idx not in graph
+                graph[idx] = {
+                    DAGNodeOperationType.READ: read_node,
+                    DAGNodeOperationType.COMPUTE: compute_node,
+                    DAGNodeOperationType.WRITE: write_node,
+                }
+
+        from ray.dag import (
```

Comment:
is it necessary to import here? can we move to the top?

Reply:
> is it necessary to import here?

Yes, it is necessary to avoid circular imports. I added a comment. https://github.com/ray-project/ray/pull/46911/commits/d7bda839575df05aa8fca0db7294168237ba0de0